<a href="https://colab.research.google.com/github/AashiDutt/AI-and-ML-for-Coders/blob/main/Chapter_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This colab contains code for:

1. CNN for Fashion MNIST dataset
2. CNN for Horses and Human classification
3. Image augmentation to enhance results for Horses and Humans classification.

In [ ]:
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

data = tf.keras.datasets.fashion_mnist

(training_images, training_labels),(test_images, test_labels) = data.load_data()

training_images = training_images.reshape(60000, 28,28,1)
training_images = training_images/255.0

test_images = test_images.reshape(10000, 28,28,1)
test_images = test_images / 255.0


#64 convolutions, 3x3 filter size, (28,28,1) - image shape= 28x28, 1 - 3rd dimension (multicolor images)
model = tf.keras.Sequential([tf.keras.layers.Conv2D(64,(3,3), activation = 'relu', input_shape = (28,28,1)), 
tf.keras.layers.MaxPooling2D(2,2),
tf.keras.layers.Conv2D(64,(3,3), activation = 'relu'),
tf.keras.layers.MaxPooling2D(2,2),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(128, activation = tf.nn.relu),
tf.keras.layers.Dense(10, activation = tf.nn.softmax)
])

model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])

model.fit(training_images, training_labels, epochs = 50)

model.evaluate(test_images, test_labels)





4431872/4422102 [==============================] - 0s 0us/step
Epoch 1/50
1875/1875 [==============================] - 61s 32ms/step - loss: 0.4398 - accuracy: 0.8396
Epoch 2/50
1875/1875 [==============================] - 60s 32ms/step - loss: 0.2936 - accuracy: 0.8913
Epoch 3/50
1875/1875 [==============================] - 60s 32ms/step - loss: 0.2489 - accuracy: 0.9076
Epoch 4/50
1875/1875 [==============================] - 59s 32ms/step - loss: 0.2166 - accuracy: 0.9189
Epoch 5/50
1875/1875 [==============================] - 60s 32ms/step - loss: 0.1911 - accuracy: 0.9278
Epoch 6/50
1875/1875 [==============================] - 59s 32ms/step - loss: 0.1661 - accuracy: 0.9377
Epoch 7/50
1875/1875 [==============================] - 59s 32ms/step - loss: 0.1468 - accuracy: 0.9443
Epoch 8/50
1875/1875 [==============================] - 59s 32ms/step - loss: 0.1300 - accuracy: 0.9510
Epoch 9/50
1875/1875 [==============================] - 59s 32ms/step - loss: 0.1139 - accuracy: 0.9559
E

[0.9172788858413696, 0.9085000157356262]

In [ ]:
classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0])



[8.0811058e-27 6.1031894e-22 6.7233276e-36 1.3615341e-24 1.7673894e-23
 3.9219445e-17 7.4316769e-32 9.2342840e-23 9.2539222e-31 1.0000000e+00]
9


In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 64)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        36928     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 flatten (Flatten)           (None, 1600)              0         
                                                                 
 dense (Dense)               (None, 128)               2

In [2]:
import urllib.request
import zipfile

In [3]:
url = "https://storage.googleapis.com/laurencemoroney-blog.appspot.com/horse-or-human.zip"

file_name = "horse-or-human.zip" #downloads zip file
training_dir = 'horse-or-human/training/' #unzips it into a directory of training data(Parent directory)
urllib.request.urlretrieve(url, file_name)

zip_ref = zipfile.ZipFile(file_name, 'r')
zip_ref.extractall(training_dir)
zip_ref.close()

Training Data - 1000+ images of 300 x 300 pixels

Validation Data - 256 images of 300 x 300 pixels

In [4]:
#To use ImageDataGenerator 

from tensorflow.keras.preprocessing.image import ImageDataGenerator

#instance for ImageDataGenerator with image augmentation
train_datagen = ImageDataGenerator(rescale = 1./ 255, rotation_range= 40,
                                        width_shift_range= 0.2, height_shift_range= 0.2,
                                        shear_range = 0.2, zoom_range = 0.2,
                                        horizontal_flip = True, fill_mode = 'nearest') 

#generating training images
train_generator = train_datagen.flow_from_directory(training_dir, 
                                                    target_size = (300,300), 
                                                    class_mode = 'binary')

Found 1027 images belonging to 2 classes.


In [5]:
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import optimizers

In [6]:
# CNN Architecture
# More layers needed - larger images (300 x 300 pixels)
# Images are colored - 3 channels (R, G, B)
# Binary classifier can be implemented with just 1 output neuron (output = 0 or 1)

model = tf.keras.models.Sequential([
          tf.keras.layers.Conv2D(16,(3,3), activation = 'relu', input_shape = (300,300, 3)),
          tf.keras.layers.MaxPooling2D(2,2), #needs no parameters - only gives max pixels to next layer
          tf.keras.layers.Conv2D(32,(3,3), activation = 'relu'),
          tf.keras.layers.MaxPooling2D(2,2),
          tf.keras.layers.Conv2D(32,(3,3), activation = 'relu'),
          tf.keras.layers.MaxPooling2D(2,2),
          tf.keras.layers.Conv2D(32,(3,3), activation = 'relu'),
          tf.keras.layers.MaxPooling2D(2,2),
          tf.keras.layers.Conv2D(32,(3,3), activation = 'relu'),
          tf.keras.layers.MaxPooling2D(2,2),
          tf.keras.layers.Flatten(),
          tf.keras.layers.Dense(512, activation = 'relu'),
          tf.keras.layers.Dense(1, activation = 'sigmoid')

])

In [7]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 298, 298, 16)      448       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 149, 149, 16)     0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 147, 147, 32)      4640      
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 73, 73, 32)       0         
 2D)                                                             
                                                                 
 conv2d_2 (Conv2D)           (None, 71, 71, 32)        9248      
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 35, 35, 32)       0

In [8]:
model.compile(loss='binary_crossentropy',
       optimizer= tf.keras.optimizers.RMSprop(lr = 0.001),
       metrics=['accuracy'])

history = model.fit(train_generator, epochs = 15)

/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/rmsprop.py:130: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(RMSprop, self).__init__(name, **kwargs)


Epoch 1/15
33/33 [==============================] - 34s 684ms/step - loss: 0.6964 - accuracy: 0.5969
Epoch 2/15
33/33 [==============================] - 23s 696ms/step - loss: 0.5414 - accuracy: 0.7176
Epoch 3/15
33/33 [==============================] - 23s 696ms/step - loss: 0.4859 - accuracy: 0.7760
Epoch 4/15
33/33 [==============================] - 23s 702ms/step - loss: 0.3562 - accuracy: 0.8306
Epoch 5/15
33/33 [==============================] - 23s 693ms/step - loss: 0.3381 - accuracy: 0.8608
Epoch 6/15
33/33 [==============================] - 23s 693ms/step - loss: 0.3079 - accuracy: 0.8870
Epoch 7/15
33/33 [==============================] - 23s 697ms/step - loss: 0.2649 - accuracy: 0.8929
Epoch 8/15
33/33 [==============================] - 23s 699ms/step - loss: 0.2221 - accuracy: 0.9299
Epoch 9/15
33/33 [==============================] - 23s 701ms/step - loss: 0.2046 - accuracy: 0.9348
Epoch 10/15
33/33 [==============================] - 23s 703ms/step - loss: 0.1532 - accura

In [9]:
# Generating Validation images
validation_url = "https://storage.googleapis.com/laurencemoroney-blog.appspot.com/validation-horse-or-human.zip"

validation_file_name = "validation-horse-or-human.zip"
validation_dir = 'horse-or-human/validation/'
urllib.request.urlretrieve(validation_url, validation_file_name)

zip_ref = zipfile.ZipFile(validation_file_name, 'r')
zip_ref.extractall(validation_dir)
zip_ref.close()


In [10]:
#ImageData Generator

validation_datagen = ImageDataGenerator(rescale=1./255)

validation_generator = train_datagen.flow_from_directory(validation_dir,
                                                         target_size= (300,300),
                                                         class_mode = 'binary')



Found 256 images belonging to 2 classes.


In [11]:
#Update history

history = model.fit(train_generator, epochs = 15,
                    validation_data = validation_generator)

Epoch 1/15
33/33 [==============================] - 29s 874ms/step - loss: 0.1117 - accuracy: 0.9533 - val_loss: 1.3538 - val_accuracy: 0.7148
Epoch 2/15
33/33 [==============================] - 28s 849ms/step - loss: 0.1203 - accuracy: 0.9679 - val_loss: 1.9356 - val_accuracy: 0.6094
Epoch 3/15
33/33 [==============================] - 28s 848ms/step - loss: 0.0787 - accuracy: 0.9757 - val_loss: 2.9299 - val_accuracy: 0.6445
Epoch 4/15
33/33 [==============================] - 28s 853ms/step - loss: 0.0731 - accuracy: 0.9688 - val_loss: 3.3962 - val_accuracy: 0.6367
Epoch 5/15
33/33 [==============================] - 28s 850ms/step - loss: 0.2054 - accuracy: 0.9611 - val_loss: 2.1839 - val_accuracy: 0.7109
Epoch 6/15
33/33 [==============================] - 28s 850ms/step - loss: 0.0657 - accuracy: 0.9776 - val_loss: 1.8801 - val_accuracy: 0.6797
Epoch 7/15
33/33 [==============================] - 28s 847ms/step - loss: 0.0968 - accuracy: 0.9698 - val_loss: 0.6901 - val_accuracy: 0.8281

In [12]:
# uploading testing images

import numpy as np
from google.colab import files
from keras.preprocessing import image

uploaded = files.upload()

for fn in uploaded.keys():

  #predicting images
  path = '/content/' + fn
  img = image.load_img(path, target_size = (300, 300)) # converts shape of input image to 300 x 300(input of model)
  x = image.img_to_array(img) # convert image to 2D array but we need 3D array
  x = np.expand_dims(x, axis = 0) # this takes care of that

  image_tensor = np.vstack([x])

  classes = model.predict(image_tensor) # model returns array containing classifications

  print(classes)
  print(classes[0])
  if classes[0] > 0.5:
    print(fn + "is a human")

  else:
    print(fn + "is a horse")

Saving girl1.jfif to girl1.jfif
Saving horse.jfif to horse.jfif
[[1.]]
[1.]
girl1.jfifis a human
[[1.]]
[1.]
horse.jfifis a human


**Before Image Augmentation**

Training accuracy - 98%

Validation Accuracy - 80%

*Model Overfitting*

**After image augmentation**

Training accuracy - 96% (decreased)

Validation Accuracy - 82% (2% increase)

Model Underfitting